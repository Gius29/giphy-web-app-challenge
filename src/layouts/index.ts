// Each constant represent a possible layout that the app may have.
export enum LayoutIDs {
  ONE_COLUMN = "ONE_COLUMN",
  THREE_COLUMNS = "THREE_COLUMNS"
}
