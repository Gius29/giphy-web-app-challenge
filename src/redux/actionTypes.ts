export enum HomePageActionType {
  DELETE_SEARCH = "DELETE_SEARCH",
  SEARCH_QUERY = "NEW_SEARCH_QUERY",
  SEARCH_QUERY_PENDING = "NEW_SEARCH_QUERY_PENDING",
  SEARCH_QUERY_FULFILLED = "NEW_SEARCH_QUERY_FULFILLED",
  SEARCH_QUERY_REJEXTED = "NEW_SEARCH_QUERY_REJEXTED",
  CHANGE_LAYOUT = "CHANGE_LAYOUT"
}
