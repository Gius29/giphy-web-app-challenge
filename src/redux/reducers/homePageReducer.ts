import { HomePageState } from "../state/HomePageState";
import { HomePageActionType } from "../actionTypes";
import { LayoutIDs } from "../../layouts";

const defaultState: HomePageState = {
  cards: [],
  query: "",
  searchOffset: 0,
  searchLimit: 10,
  searchResultPending: false,
  layoutID: LayoutIDs.ONE_COLUMN
};

export const homePageReducer = (
  state: HomePageState = defaultState,
  action: any
) => {
  switch (action.type) {
    case HomePageActionType.DELETE_SEARCH:
    // [IMPROVMENT] The error should display a meaningful message to the user.
    case HomePageActionType.SEARCH_QUERY_REJEXTED:
      return { ...defaultState, layoutID: state.layoutID };

    case HomePageActionType.SEARCH_QUERY_PENDING:
      return { ...state, searchResultPending: true };

    case HomePageActionType.SEARCH_QUERY_FULFILLED:
      // Increasing the pagination offeset value to have it ready for next api call
      let newSearchOffset = state.searchOffset + state.searchLimit;

      if (action.payload.items.length === 0) {
        //The negative number indicates that the list is ended
        newSearchOffset = -1;
      }

      // Extracting from the API response the content that is needed to create a card (CardModel)
      const newCards = action.payload.items.map((item: any) => ({
        id: item.id, // The id should be unique, if not we need to add a salt
        bitlyUrl: item.bitly_url,
        gifUrl: item.images.original.url,
        gifAlt: item.title || "gif image"
      }));

      return {
        ...state,
        query: action.payload.query,
        searchOffset: newSearchOffset,
        searchResultPending: false,
        cards: [...state.cards, ...newCards]
      };

    case HomePageActionType.CHANGE_LAYOUT:
      let newLayout = LayoutIDs.THREE_COLUMNS;
      if (state.layoutID === LayoutIDs.THREE_COLUMNS) {
        newLayout = LayoutIDs.ONE_COLUMN;
      }
      return { ...state, layoutID: newLayout };

    default:
      return state;
  }
};
