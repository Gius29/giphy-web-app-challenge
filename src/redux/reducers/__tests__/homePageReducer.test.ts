import { LayoutIDs } from "../../../layouts";
import { HomePageActionType } from "../../actionTypes";
import { homePageReducer } from "../homePageReducer";

describe("homePageReducer", () => {
  const defaultState: HomePageState = {
    cards: [],
    query: "",
    searchOffset: 0,
    searchLimit: 10,
    searchResultPending: false,
    layoutID: LayoutIDs.ONE_COLUMN
  };

  describe("should return the deafult state", () => {
    it("with unknown action type", () => {
      const result = homePageReducer(undefined, { type: "STRANGE TYPE" });

      expect(result).toEqual(defaultState);
    });

    it("with DELETE_SEARCH", () => {
      const result = homePageReducer(undefined, {
        type: HomePageActionType.DELETE_SEARCH
      });

      expect(result).toEqual(defaultState);
    });

    it("with DELETE_SEARCH", () => {
      const result = homePageReducer(undefined, {
        type: HomePageActionType.SEARCH_QUERY_REJEXTED
      });

      expect(result).toEqual(defaultState);
    });
  });

  it("should set searchResultPending to true with SEARCH_QUERY_PENDING", () => {
    const result = homePageReducer(undefined, {
      type: HomePageActionType.SEARCH_QUERY_PENDING
    });

    expect(result).toEqual({ ...defaultState, searchResultPending: true });
  });

  it("should set right layoutID with CHANGE_LAYOUT", () => {
    // Starting with layout: LayoutIDs.THREE_COLUMNS
    let result = homePageReducer(
      { ...defaultState },
      {
        type: HomePageActionType.CHANGE_LAYOUT
      }
    );

    expect(result.layoutID).toEqual(LayoutIDs.THREE_COLUMNS);

    // Starting with layout: LayoutIDs.ONE_COLUMN
    result = homePageReducer(
      { ...defaultState, layoutID: LayoutIDs.THREE_COLUMNS },
      {
        type: HomePageActionType.CHANGE_LAYOUT
      }
    );

    expect(result.layoutID).toEqual(LayoutIDs.ONE_COLUMN);
  });

  describe("SEARCH_QUERY_FULFILLED", () => {
    it("should update the state with items in the response", () => {
      const result = homePageReducer(
        { ...defaultState },
        {
          type: HomePageActionType.SEARCH_QUERY_FULFILLED,
          payload: {
            items: [
              {
                id: "one",
                bitly_url: "bitly_url1",
                images: {
                  original: { url: "url1" }
                },
                title: "gifAlt1"
              },
              {
                id: "two",
                bitly_url: "bitly_url2",
                images: {
                  original: { url: "url2" }
                }
              }
            ]
          }
        }
      );

      expect(result).toMatchSnapshot();
    });

    it("should update the state with no items in the response", () => {
      const result = homePageReducer(
        { ...defaultState },
        {
          type: HomePageActionType.SEARCH_QUERY_FULFILLED,
          payload: {
            items: []
          }
        }
      );

      expect(result).toMatchSnapshot();
    });
  });
});
