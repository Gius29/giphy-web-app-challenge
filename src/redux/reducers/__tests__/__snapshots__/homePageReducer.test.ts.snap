// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homePageReducer SEARCH_QUERY_FULFILLED should update the state with items in the response 1`] = `
Object {
  "cards": Array [
    Object {
      "bitlyUrl": "bitly_url1",
      "gifAlt": "gifAlt1",
      "gifUrl": "url1",
      "id": "one",
    },
    Object {
      "bitlyUrl": "bitly_url2",
      "gifAlt": "gif image",
      "gifUrl": "url2",
      "id": "two",
    },
  ],
  "layoutID": "ONE_COLUMN",
  "query": undefined,
  "searchLimit": 10,
  "searchOffset": 10,
  "searchResultPending": false,
}
`;

exports[`homePageReducer SEARCH_QUERY_FULFILLED should update the state with no items in the response 1`] = `
Object {
  "cards": Array [],
  "layoutID": "ONE_COLUMN",
  "query": undefined,
  "searchLimit": 10,
  "searchOffset": -1,
  "searchResultPending": false,
}
`;
